src/ecs/resources/input/input_handler.rs

```diff
@@ -0,0 +1,329 @@
+//! World resource that handles all user input.
+
+use engine::{ElementState, WindowEvent, Event, VirtualKeyCode, MouseButton};
+use smallvec::SmallVec;
+use super::*;
+
+/// This struct holds state information about input devices.
+///
+/// For example, if a key is pressed on the keyboard, this struct will record
+/// that the key is pressed until it is released again. Usage requires pumping input events through
+/// this struct, using the following code in the [handle_events] method of [State].
+///
+/// [State]: ../../../engine/trait.State.html
+/// [handle_events]: ../../../engine/trait.State.html#method.handle_events
+///
+/// ```ignore
+/// fn handle_events(&mut self,
+///                  events: &[WindowEvent],
+///                  _: &mut World,
+///                  _: &mut AssetManager,
+///                  _: &mut Pipeline)
+///                  -> Trans {
+///     // ...
+///     let mut input_handler = world.write_resource::<InputHandler>();
+///     input_handler.update(events);
+///     // ...
+/// }
+/// ```
+#[derive(Default)]
+pub struct InputHandler {
+    /// Maps inputs to actions and axes.
+    pub bindings: Bindings,
+    pressed_keys: SmallVec<[VirtualKeyCode; 16]>,
+    down_keys: SmallVec<[VirtualKeyCode; 16]>,
+    released_keys: SmallVec<[VirtualKeyCode; 16]>,
+    pressed_mouse_buttons: SmallVec<[MouseButton; 16]>,
+    down_mouse_buttons: SmallVec<[MouseButton; 16]>,
+    released_mouse_buttons: SmallVec<[MouseButton; 16]>,
+    mouse_position: Option<(i32, i32)>,
+    previous_mouse_position: Option<(i32, i32)>,
+    text_this_frame: String,
+}
+
+impl InputHandler {
+    /// Creates a new input handler.
+    pub fn new() -> InputHandler {
+        InputHandler {
+            pressed_keys: SmallVec::new(),
+            down_keys: SmallVec::new(),
+            released_keys: SmallVec::new(),
+            pressed_mouse_buttons: SmallVec::new(),
+            down_mouse_buttons: SmallVec::new(),
+            released_mouse_buttons: SmallVec::new(),
+            mouse_position: None,
+            previous_mouse_position: None,
+            bindings: Bindings::default(),
+            text_this_frame: String::new(),
+        }
+    }
+
+    /// Updates the input handler with new engine events.
+    pub fn update(&mut self, events: &[WindowEvent]) {
+        // Before processing these events clear the single frame vectors
+        self.down_keys.clear();
+        self.released_keys.clear();
+        self.down_mouse_buttons.clear();
+        self.released_mouse_buttons.clear();
+        self.previous_mouse_position = self.mouse_position;
+        self.text_this_frame.clear();
+        for event in events {
+            match event.payload {
+                Event::ReceivedCharacter(c) => {
+                    self.text_this_frame.push(c);
+                }
+                Event::KeyboardInput(ElementState::Pressed, _, Some(key_code)) => {
+                    if self.pressed_keys.iter().all(|&k| k != key_code) {
```

Comment:
@omni-viral Additional pressed signals are sent by some backends when the key "repeats" due to being held down, it's very important to the remove code that any entry be present once and only once, so this keeps the repeat signals from getting added in.